### PR TITLE
CAT-2598 Refactor FormulaEditorEditTextTest

### PR DESCRIPTION
**Pull request by AntiDog**
removed ugly stuff

### CODE REVIEW

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+		onBrickAtPosition(1)
+				.checkShowsText("0 ");
+	}
+
+	@Category({Cat.CatrobatLanguage.class, Level.Smoke.class})
+	@Test
+	public void testDiscardDialog() {
+		onFormulaEditor()
+				.performEnterFormula("99");
+		pressBack();
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java]: please split this test into smaller tests. its really hard to follow whats going on.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+		onView(withText(getCurrentActivity().getString(R.string.formula_editor_discard_changes_dialog_title)))
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java]: "withText" also takes string ids:

onView(withText(R.string.formula_editor_discard_changes_dialog_title))

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+		onFormulaEditor()
+				.performEnterFormula(".9");
+		pressBack();
+		onView(withText(getCurrentActivity().getString(R.string.formula_editor_discard_changes_dialog_title)))
+				.check(matches(isDisplayed()));
+		onView(withText(yes))
+				.perform(click());
+		checkToastMessage(R.string.formula_editor_changes_saved);
+
+		onView(allOf(withId(R.id.brick_change_size_by_edit_text), withEffectiveVisibility(Visibility.VISIBLE)))
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java]: please access bricks via the wrapper methods. (eg. onBrickAtPosition... when in script view).
due to the pressBack() switching around its really hard to follow... 
the onBrickAtPosition would give the additional info that now were back in the scriptFragment

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+				.perform(click());
+		pressBack();
+
+		onView(allOf(withId(R.id.brick_change_size_by_edit_text), withEffectiveVisibility(Visibility.VISIBLE)))
+				.check(matches(withText("99" + decimalMark + "9 ")));
+
+		onView(allOf(withId(R.id.brick_change_size_by_edit_text), withEffectiveVisibility(Visibility.VISIBLE)))
+				.perform(click());
+		onView(withId(R.id.formula_editor_keyboard_delete))
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java]: please use only wrapper methods of the formula editor wrapper instead of using ids directly.

formula editor will be replaced / refactored, so if all the ids change it would be nice that we could only update a single wrapper instead of a s*load of tests that reference every single button on it...

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+				.perform(click());
+		onView(withId(R.id.formula_editor_keyboard_compute))
+				.perform(click());
+		onView(withId(R.id.formula_editor_compute_dialog_textview))
+				.check(matches(withText("-8.111")));
+	}
+
+	private int getAbsoluteCursorPosition() {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java]: it would be even nicer if youd put this in a custom matcher.
and then add this to the formula editor wrapper

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+		FormulaEditorEditText editText = (FormulaEditorEditText) getCurrentActivity().findViewById(R.id
+				.formula_editor_edit_field);
+		Object position = Reflection.getPrivateField(editText, "absoluteCursorPosition");
+		assertTrue("Cursor position not valid", position instanceof Number);
+		return ((Number) position).intValue();
+	}
+
+	private void setAbsoluteCursorPosition(int position) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java]: same as above, just a custom action instead of a custom matcher

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java (hunk 5 above)
Comment (AntiDog) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java]: do you mean I should add a function like this:
```
public FormulaEditorWrapper checkHasCursorPosition(int expected)
{
    FormulaEditorEditText editText = (FormulaEditorEditText) getCurrentActivity().findViewById(R.id.formula_editor_edit_field);
    Object position = Reflection.getPrivateField(editText, "absoluteCursorPosition");
    assertEquals(((Number) position).intValue(),expected);
    return new FormulaEditorWrapper();
} 
```
to the FormulaEditorWrapper? This seems somehow wrong to me since I directly access private fields of the FormulaEditorEditText and this isn't really Espresso stuff and I have no idea how to do this in Espresso or if this is even possible to do

**Comment by AntiDog** [comment]
Tried to apply suggested changes but I'm unsure if you really meant I should change setAbsoluteCursorPosition & checkHasCursorPosition to this. But I have no idea how to do it otherwise?

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/utils/FormulaEditorWrapper.java [hunk 7]
```diff
@@ -109,6 +115,24 @@ public FormulaEditorWrapper checkShows(String expected) {
 		return new FormulaEditorWrapper();
 	}
 
+	public FormulaEditorWrapper checkHasCursorPosition(int expected) {
+		FormulaEditorEditText editText = (FormulaEditorEditText) getCurrentActivity()
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/utils/FormulaEditorWrapper.java]: just saw this now. getCurrentActivity is not safe to be used like this. this is what custom matchers are intended for, so you dont have to aquire the activity like this.

**Comment by 84n4n4** [comment]
checkHasCursorPosition and setAbsoluteCursorPosition should be changed, implement a new viewMatcher for the checkHasCursorPosition and call it something like cursorAt(int position), then it can be used as
.check(matches(cursorAt(5)))
and then put this functionality in the existing formula editor wrapper.
the setAbsoluteCursorPosition should be implemented as a custom action, and also moved to the formulaeditor wrapper.
there are a few actions already in the codebase you can have a look at, just search for "implements ViewAction" and "static ViewAction" in the uiespresso package, or have a look at examples online. same with the matchers, theres a lot of them around already in the uiespresso package, eg have a look at RecyclerViewItemMatcher to see how to properly aquire an ui object without getting the activity first...

**Comment by AntiDog** [comment]
@84n4n4 are you sure this is possible by doing it with a custom action? I'm running into strange issues while trying to do so. I think I did it right because the action does set the cursor, however, it isn't "remembered" once the some of the keyboard buttons are pressed. Could you please take a look this album which is containing screenshots of the issue and is explaining the details: https://imgur.com/a/HYP83

It didn't happen with the old implementation and only setting the position seems to work, if I check it right after setting the position everything is as it should be (visually and via code).

Did I mess something up with the custom Action? I never did one before so it could very well be my fault... 

(All this doesn't happen if you click manually at the position)

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+	@Rule
+	public BaseActivityInstrumentationRule<SpriteActivity> baseActivityTestRule = new
+			BaseActivityInstrumentationRule<>(SpriteActivity.class, true, false);
+
+	@Before
+	public void setUp() throws Exception {
+		Script script = BrickTestUtils.createProjectAndGetStartScript("FormulaEditorEditTextTest");
+		script.addBrick(new ChangeSizeByNBrick());
+		baseActivityTestRule.launchActivity(null);
+		sin = getCurrentActivity().getString(R.string.formula_editor_function_sin);
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java]: please do not use getCurrentActivity, this should be only used for asserting that one is in a certain activity and a few other rare cases.
either get the targetContext (proper) from the instrumentation, or get the activity from the rule (kind of proper), or just use the getResourcesString method from utils that uses the targetContext...

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+
+	private String sin;
+	private String random;
+	private String decimalMark;
+	private String no;
+	private String yes;
+
+	@Rule
+	public BaseActivityInstrumentationRule<SpriteActivity> baseActivityTestRule = new
+			BaseActivityInstrumentationRule<>(SpriteActivity.class, true, false);
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java]: please use 
BaseActivityInstrumentationRule<>(SpriteActivity.class, SpriteActivity.EXTRA_FRAGMENT_POSITION,
			SpriteActivity.FRAGMENT_SCRIPTS)

and baseActivityTestRule.launchActivity()  (without the null)

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+		yes = getCurrentActivity().getString(R.string.yes);
+		decimalMark = getCurrentActivity().getString(R.string.formula_editor_decimal_mark);
+		onView(withId(R.id.brick_change_size_by_edit_text))
+				.perform(click());
+	}
+
+	@Category({Cat.CatrobatLanguage.class, Level.Smoke.class})
+	@Test
+	public void testSingleTapOnFunctionName() {
+		onView(withId(R.id.formula_editor_keyboard_function))
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java]: theres wrapper methods for selectiong formulas, please use those, i do not want to have any ids of formula editor elements in the tests

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.uiespresso.content.brick.app;
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java]: please move this test to the formulaEditor package.

Comment (AntiDog) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java]: Ok sure. Will try to fix these issues ASAP.
Have you already taken a look at my question?
 (https://github.com/Catrobat/Catroid/pull/2724#issuecomment-368007494)
I already tried several things but none of them worked. I really need some help with that issue I tried everything I could come up with by now... Would you mind taking a look into this together some time next week in the panic room?

**Comment by 84n4n4** [comment]
ok, i had a closer look on whats going on in formulaEditorEditText to set the position.
now i wonder how this could ever work (and not why its not working anymore in the action).
i found a simpler way to match for the position, so you can just use  getSelectionStart on the editText, but the other way round for setting a specific selection is pretty much not possible the way this was implemented. (please go ahead and have a look at FormulaEditorEditText on how to not implement an edit text and fuck it up for everyone else.) this is fubar.
so i propose dropping the tests that set a cursor position.

**Comment by AntiDog** [comment]
OK

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/utils/FormulaEditorWrapper.java [hunk 12]
```diff
@@ -130,6 +132,16 @@ public void performClickOn(Matcher<View> matcher) {
 				.perform(click());
 	}
 
+	public void performLongClickOn(Matcher<View> matcher) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/utils/FormulaEditorWrapper.java]: since theres only one usage for each of those two methods, and they are rather specific for those tests they should not be in this wrapper (nobody else is going to need those, and theres little added readability)

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextTest.java [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+		no = getResourcesString(R.string.no);
+		yes = getResourcesString(R.string.yes);
+		decimalMark = getResourcesString(R.string.formula_editor_decimal_mark);
+		onBrickAtPosition(1)
+				.perform(click());
+	}
+
+	@Category({Cat.CatrobatLanguage.class, Level.Smoke.class})
+	@Test
+	public void testHighlightColor() {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextTest.java]: please remove this test

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextGenericTests.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+				{R.id.formula_editor_keyboard_1, "1 ", "button_1"},
+				{R.id.formula_editor_keyboard_2, "2 ", "button_2"},
+				{R.id.formula_editor_keyboard_3, "3 ", "button_3"},
+				{R.id.formula_editor_keyboard_4, "4 ", "button_4"},
+				{R.id.formula_editor_keyboard_5, "5 ", "button_5"},
+				{R.id.formula_editor_keyboard_6, "6 ", "button_6"},
+				{R.id.formula_editor_keyboard_7, "7 ", "button_7"},
+				{R.id.formula_editor_keyboard_8, "8 ", "button_8"},
+				{R.id.formula_editor_keyboard_9, "9 ", "button_9"},
+				{R.id.formula_editor_keyboard_decimal_mark, "0" + getResources().getString(R.string
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextGenericTests.java]: theres also getResourcesString in uitestutils, which to further increase readability you could proxy through a new method in this test to eg. getString().
if you add the " " in the test where you check them it would clean this up significantly...

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextGenericTests.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+		script.addBrick(new ChangeSizeByNBrick());
+		baseActivityTestRule.launchActivity(null);
+		onView(withId(R.id.brick_change_size_by_edit_text))
+				.perform(click());
+	}
+
+	@Parameterized.Parameters(name = "{2}" + "-Test")
+	public static Iterable<Object[]> data() {
+		return Arrays.asList(new Object[][] {
+				{R.id.formula_editor_keyboard_0, "0 ", "button_0"},
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextGenericTests.java]: please use the matchers already provided in formulaEditorWrapper (NUM0, NUM1, etc..) or even go there and just have one parameter for the test, eg. "8"
and then enter it via performEnterFormula or performEnterNumber and then check if it is in the textfield with an added " " at the end.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextGenericTests.java [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+import static org.catrobat.catroid.uiespresso.formulaeditor.utils.FormulaEditorWrapper.onFormulaEditor;
+import static org.catrobat.catroid.uiespresso.util.UiTestUtils.getResources;
+
+@Category({Cat.AppUi.class, Level.Smoke.class})
+@RunWith(Parameterized.class)
+public class FormulaEditorEditTextGenericTests {
+
+	@Rule
+	public BaseActivityInstrumentationRule<SpriteActivity> baseActivityTestRule = new
+			BaseActivityInstrumentationRule<>(SpriteActivity.class, true, false);
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextGenericTests.java]: please state the fragment you want to launch in the rules constructor (already updated in all other tests in code base), and the launchActivity() (without the null)

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextGenericTests.java (hunk 14 above)
Comment (AntiDog) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextGenericTests.java]: I don't really get what you mean by "proxying through a new method" and using getResourcesString doesn't improve anything and IMO it makes it even worse since android studio already has the option to only show the text "." instead of getResources().getString(...) while it shows getResourcesString(".") if I use the function. Please see the screenshot below if this explanation isn't clear :S
Difference between the two scenarios

For me, it doesn't make sense to use getResourcesString() over the getResources().getString() since it makes it harder to read because android studio doesn't know that function and hence it can't view it simplified. However, if you really want it this way I can, of course, change it to this I'm just unsure if I understand correctly what you mean...

**Comment by 84n4n4** [comment]
test this please

**Comment by 84n4n4** [comment]
please run espresso tests

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextTest.java [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+		onFormulaEditor()
+				.checkShows(" ");
+	}
+
+	@Category({Cat.CatrobatLanguage.class, Level.Smoke.class})
+	@Test
+	public void testFunctionDeletion() {
+		onFormulaEditor()
+				.performOpenCategory(FUNCTIONS);
+		onView(withText(containsString(random)))
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextTest.java]: please use:
onFormulaEditor()
				.performOpenCategory(FormulaEditorWrapper.Category.FUNCTIONS)
				.performSelect(random);

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextTest.java [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+				.perform(click());
+		onFormulaEditor()
+				.performClickOn(BACKSPACE);
+		onFormulaEditor()
+				.performClickOn(BACKSPACE);
+
+		onFormulaEditor()
+				.performOpenCategory(FUNCTIONS);
+
+		onView(allOf(withText(containsString(random)), withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE)))
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextTest.java]: onFormulaEditor()
				.performOpenCategory(FormulaEditorWrapper.Category.FUNCTIONS)
				.performSelect(random);

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextTest.java [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+		onFormulaEditor()
+				.performClickOn(BACKSPACE);
+		onFormulaEditor()
+				.performClickOn(COMPUTE);
+		onView(withId(R.id.formula_editor_compute_dialog_textview))
+				.check(matches(withText("-8.111")));
+	}
+
+	// https://stackoverflow.com/questions/28390574/checking-toast-message-in-android-espresso
+	public void checkToastMessage(int messageID) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextTest.java]: i really dislike "check..." methods, since it is never clear what it actually does.
either rename this to "assertToastMessageDisplayed", or even better, have this function return the view matcher and rename to onToastMessage:
ViewInteraction onToastMessage(ViewMatcher matcher) {
	return onView(someMatcher)
			.inRoot(withDecorView(Matchers.not(baseActivityTestRule.getActivity().getWindow()
			.getDecorView())));
}

and then use it as:
onToastMessage(withText(messageId))
	.check(matches(isDisplayed));

this would have the advantage that you can pass anything as a matcher, and since withText() works with Ids as well as strings you get both for free.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/utils/FormulaEditorWrapper.java [hunk 20]
```diff
@@ -130,6 +131,11 @@ public void performClickOn(Matcher<View> matcher) {
 				.perform(click());
 	}
 
+	public void performDoubleClickOn(Matcher<View> matcher) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/utils/FormulaEditorWrapper.java]: this is so specific to a single test that it doesnt have any business in here. please move to the test, (but use the static matcher from the wrapper)

**Comment by 84n4n4** [comment]
test this please

**Comment by 84n4n4** [comment]
theres now a toast matcher:
	public static ViewInteraction onToast(Matcher<View> viewMatcher) {


to be used like this:
onToast(withText("some string or id")).check(matches(isDisplayed))

**Comment by 84n4n4** [comment]
test this please

**Comment by 84n4n4** [comment]
please run espresso tests

**Comment by 84n4n4** [comment]
please run espresso tests
